### PR TITLE
Bdev zero with discard

### PR DESCRIPTION
**Pull request by majianpeng**
add function make bluestore support discard. If block device support discard-zeroes-data which after discard the data are zero, we use discard instead of write-zero.

### CODE REVIEW

**Comment by majianpeng** [comment]
@yuyuyu101 . Have you time to review this? Thanks!

File: src/os/bluestore/KernelDevice.cc [hunk 1]
```diff
@@ -127,6 +129,19 @@ int KernelDevice::open(string p)
...
   assert(r == 0);
 
+  //check disk whether support discard  & discard zero data
+  if (S_ISBLK(st.st_mode)) {
+    char realname[PATH_MAX] = {0};
+    if (readlink(path.c_str(), realname, sizeof(realname) - 1) != -1) {
+      discard = get_block_device_int_property(realname, "discard_granularity");
+      if (discard)
+	dout(1) << realname << " support discard command"  << dendl;
+      discard_zeroes_data = get_block_device_int_property(realname, "discard_zeroes_data");
```
Comment (yuyuyu101) [inline src/os/bluestore/KernelDevice.cc]: actually I think this way may more elegant:
if (ioctl(s->fd, BLKDISCARDZEROES, &arg) == 0 && arg) {
    discard_zeroes_data = true;
}

although get_block_device_int_property should be ok too

File: src/os/bluestore/KernelDevice.cc [hunk 2]
```diff
@@ -127,6 +129,19 @@ int KernelDevice::open(string p)
...
+      if (discard_zeroes_data)
+	dout(1) << realname << " support discard zero data" << dendl;
```
Comment (yuyuyu101) [inline src/os/bluestore/KernelDevice.cc]: /* On Linux 3.10, BLKDISCARD leaves stale data in the page cache.  Do
 * not rely on the contents of discarded blocks unless using O_DIRECT.
 * Same for BLKZEROOUT.

from qemu source code

File: src/os/bluestore/KernelDevice.cc (hunk 1 above)
Comment (majianpeng) [inline src/os/bluestore/KernelDevice.cc]: @yuyuyu101  different os implement different get_block_device_int_property. From this point, i'd   think my method is better.

File: src/os/bluestore/KernelDevice.cc (hunk 2 above)
Comment (majianpeng) [inline src/os/bluestore/KernelDevice.cc]: @yuyuyu101. For data correctly, it must remove this. Have you method to remove page cache(invalidate_cache)?

**Comment by majianpeng** [comment]
@yuyuyu101 . You mentioned discard bypass page cache which make read old data when use discard instead of write-zero. I searched kernel and don't find any interface which can invalidate page cache.
If we first call sync_file_range and lather posix_fadvise(offset, len, DONTNEED) make the kernel free the related page. I suppose the most case no page cache for those data. How about that?
@liewegas , Could you give some suggestions? 

BTW, KernelDevice::invalidate_cache use posix_fadivse(DONTNEED) to invalidate page cache. From the kernel, i think this don't make sure pages released.

**Comment by yuyuyu101** [comment]
@majianpeng yes, it should be no possible to enforce flush pagecache in userside

**Comment by majianpeng** [comment]
@yuyuyu101 . Are you mean sync_file_range? or posix_fadvise(DONTNEED)?. I think sync_file_range can do this.

**Comment by yuyuyu101** [comment]
posix_advise won't flush page if the pages are dirty. sync_file_range can flush pages if they are dirty. So I'm not whether sync_file_range will drop cache if they are clean?

**Comment by majianpeng** [comment]
I mean to drop page cache
a)sync_file_range: make page cache clean
b)posix_fadvise(DONTNEED): remove page cache.

**Comment by liewegas** [comment]
On Thu, 3 Mar 2016, majianpeng wrote:
> I mean to drop page cache
> a)sync_file_range: make page cache clean
> b)posix_fadvise(DONTNEED): remove page cache.

Right.

**Comment by majianpeng** [comment]
@liewegas . Have you time to review this? Thanks!

**Comment by majianpeng** [comment]
@liewegas @yuyuyu101 .  For ceph, before we do aio_zero, the related page cache(if existed) must clean.After a transaction, every write data flush into disk so the page cache become clean.  So we can only use posix_fadvise(DONT_NEED) to free page cache. Remove the useless sync_file_range(though it's only a syscall).  Is it right?

**Comment by majianpeng** [comment]
updated.

File: src/os/bluestore/BlockDevice.h [hunk 3]
```diff
@@ -103,6 +103,9 @@ class BlockDevice {
   virtual int invalidate_cache(uint64_t off, uint64_t len) = 0;
   virtual int open(string path) = 0;
   virtual void close() = 0;
+
+  virtual bool support_discard() = 0;
```
Comment (liewegas) [inline src/os/bluestore/BlockDevice.h]: supports_discard or can_discard

File: src/os/bluestore/BlockDevice.h [hunk 4]
```diff
@@ -103,6 +103,9 @@ class BlockDevice {
...
+  virtual int do_discard(uint64_t off, uint64_t len) = 0;
```
Comment (liewegas) [inline src/os/bluestore/BlockDevice.h]: discard

**Comment by liewegas** [comment]
@majianpeng Hmm, you know what I think we are missing and would be really helpful is code in blkdev.cc that will take a random path (say, /var/lib/ceph/osd/ceph-213/block) and map it back to a device name that can be looked up in /sys/block to get device parameters (like 'queue/rotating', to detect an ssd).  Then we can do a bunch of conditional defaults differently for HDD and SSD (like min_alloc_size).

(This would be a separate PR!)

**Comment by liewegas** [comment]
Needs rebase!

**Comment by majianpeng** [comment]
update

**Comment by yuyuyu101** [comment]
the tricky sync with sync_file_range would be a concern. I don't have much sense on this. We may enable a qa test for "bdev_discard_on_zero=true". Otherwise, we always don't know is this make sense..

**Comment by majianpeng** [comment]
Why add bluestore_discard_on_release & bdev_discard_on_zero? This because Sage said: discard isn't good. But from ext4, it use discard when free blocks. 
I want to set true for the default value and add perfcount to record the perfmance which judge whether use those feature. @liewegas, how about this?  Maybe not all ssd have discard performance performance.

At for sync_file_range, i don't know better method to implement.

**Comment by liewegas** [comment]
This needs to be recast, I think.  We no longer zero anything.  We might want to issue discard to the block device, though, when regions get freed up.  This would probably be done by the allocator or by the freelist... but I'm not sure off the top of my head how to do it efficiently.  My guess is:

- we want a BlockDevice discard operation
- we want the Allocator to give us discard op(s) when we release extents.  We have to be careful here, though, that we don't race with new allocations...
- any discards need to respect the discard granularity, and
- we need to make sure we issue the discard on the last allocation within a discard-sized block.

**Comment by liewegas** [comment]
ripping out aio_zero

**Pull request closed without merging**